### PR TITLE
Add basic integration tests for building all books

### PR DESCRIPTION
**Pull request by nik9000**
This adds integration testing for building multiple books out of a
single repository and building a single book out of multiple
repositories. To do that it significantly reworks the spec helpers
around integration testing.

### CODE REVIEW

File: integtest/spec/all_books_spec.rb [hunk 1]
```diff
@@ -1,55 +1,100 @@
...
+      ASCIIDOC
+      repo.write 'second/index.asciidoc', <<~ASCIIDOC
         = Title
 
         == Chapter
 
         Some text.
       ASCIIDOC
-      src.init_repo 'source'
-      src.write 'conf.yaml', <<~YAML
```
Comment (nik9000) [inline integtest/spec/all_books_spec.rb]: This has been moved to `Source`.

File: integtest/spec/helper/convert.rb [hunk 2]
```diff
@@ -1,44 +0,0 @@
-# frozen_string_literal: true
-
-require 'open3'
-
-##
-# Method to convert asciidoc files to html in example blocks like `it`
```
Comment (nik9000) [inline integtest/spec/helper/convert.rb]: I've moved this into `Dest.rb`

File: integtest/spec/helper/dsl/source.rb [hunk 3]
```diff
@@ -1,50 +0,0 @@
-# frozen_string_literal: true
-
-require 'open3'
-
-require_relative '../sh'
-
-##
-# Helper class for setting up source files for tests.
-module Dsl
```
Comment (nik9000) [inline integtest/spec/helper/dsl/source.rb]: I've moved this.

**Comment by nik9000** [comment]
The tests this adds are a hole in the coverage in the integration test `Makefile`. It has some more advanced cases but not these.

Those more advanced cases will need the kinds of things that I'm building here.

**Comment by nik9000** [comment]
@estolfo, could you have a look at this when you get a chance? I think it is just about that last of my "reworking integration testing" blockers. Once I get this I can work on things like #833 and #823, but without it I'm kind of stuck using `make` to test them which is a sad thing.

**Comment by nik9000** [comment]
Because this is "only" tests and it is blocking a bunch of work I've self merged this. I'd still love a review and commit to cleaning up whatever that process finds!

File: integtest/spec/helper/dest.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    out = run_convert cmd
+    sh "git clone #{bare_repo} #{@dest}"
+    out
+  end
+
+  private
+
+  ##
+  # The location of the bare repository. the first time this is called in a
+  # given contextw the bare repository is initialized
```
Comment (estolfo) [inline integtest/spec/helper/dest.rb]: `contextw` -> `context`

File: integtest/spec/helper/source.rb [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    @root = File.expand_path 'src', tmp
+    Dir.mkdir @root
+    @repos = []
+    @books = []
+  end
+
+  ##
+  # Create a source repo to which you can add files to be converted.
+  def repo(name)
+    Repo.new(name, path(name)).tap { |r| @repos.push r }
```
Comment (estolfo) [inline integtest/spec/helper/source.rb]: why would you just write
`@repos << Repo.new(name, path(name))` ?

File: integtest/spec/helper/source.rb [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+  ##
+  # Transform path fragment for a source file into the path that that file
+  # should have.
+  def path(source_relative_path)
+    File.expand_path(source_relative_path, @root)
+  end
+
+  def book(title, prefix)
+    Book.new(title, prefix).tap { |b| @books.push b }
```
Comment (estolfo) [inline integtest/spec/helper/source.rb]: same here, you could write:
`@books << Book.new(title, prefix)`

File: integtest/spec/helper/source.rb (hunk 5 above)
Comment (nik9000) [reply integtest/spec/helper/source.rb]: :man_facepalming: I think it made more sense in a previous iteration....

Comment (nik9000) [reply integtest/spec/helper/source.rb]: It turns out I need to return the new repo as well as add it which is why I did this. Could you suggest something that makes that more obvious? I'll push some comments too.

Comment (estolfo) [reply integtest/spec/helper/source.rb]: The three ways that come to mind are:
```ruby
@repos << Repo.new(name, path(name)) && @repos[-1]
```

```ruby
repo = Repo.new(name, path(name))
(@repos << repo) && repo
```

```ruby
repo = Repo.new(name, path(name))
@repos << repo
repo
```

The use of `#tap` is certainly valid but seems not that necessary.

**Pull request merged**